### PR TITLE
Nitzan/eco 141/offer caps

### PR DESCRIPTION
**Pull request by nitzantomer**

### CODE REVIEW

File: scripts/src/public/services/offers.ts [hunk 1]
```diff
@@ -46,10 +49,20 @@ function offerDbToApi(offer: db.Offer, content: db.OfferContent) {
 	};
 }
 
-async function filterOffers(offers: db.Offer[], logger: LoggerInstance): Promise<Offer[]> {
+async function filterOffers(userId: string, offers: db.Offer[], logger: LoggerInstance): Promise<Offer[]> {
 	return await Promise.all(
 		offers
 			.map(async offer => {
+				const total = await dbOrder.Order.count({ where: { offerId: offer.id } });
+				if (total === offer.cap.total) {
```
Comment (doodyparizada) [inline scripts/src/public/services/offers.ts]: I would check >=

File: scripts/src/public/services/offers.ts [hunk 2]
```diff
@@ -46,10 +49,20 @@ function offerDbToApi(offer: db.Offer, content: db.OfferContent) {
...
+					return null;
+				}
+
+				const forUser = await dbOrder.Order.count({ where: { offerId: offer.id, userId } });
+				if (forUser === offer.cap.per_user) {
```
Comment (doodyparizada) [inline scripts/src/public/services/offers.ts]: I would check >=

File: scripts/src/public/services/orders.ts [hunk 3]
```diff
@@ -63,103 +69,114 @@ function orderDbToApi(order: db.Order, logger: LoggerInstance): Order {
...
+			const forUser = await db.Order.count({
+				where: {
+					userId,
+					offerId
+				}
+			});
+
+			if (forUser === offer.cap.per_user) {
+				return undefined;
+			}
```
Comment (doodyparizada) [inline scripts/src/public/services/orders.ts]: This logic can be shared with getOffers

Comment (doodyparizada) [reply scripts/src/public/services/orders.ts]: plus, in getOffers you return `null` and here you return `undefined`

File: scripts/src/public/services/orders.ts [hunk 4]
```diff
@@ -63,103 +69,114 @@ function orderDbToApi(order: db.Order, logger: LoggerInstance): Order {
...
+				type: offer.type,
+				status: "opened",
+				meta: offer.meta.order_meta
+			});
+			await order.save();
+			return order;
+		};
+
+		// order = await lock(createOrderResourceId, fn());
+		order = await fn();
```
Comment (doodyparizada) [inline scripts/src/public/services/orders.ts]: if you're putting this in a function, give it a name?

File: scripts/src/public/services/orders.ts [hunk 5]
```diff
@@ -63,103 +69,114 @@ function orderDbToApi(order: db.Order, logger: LoggerInstance): Order {
...
 	};
 }
 
 export async function submitOrder(
 	orderId: string, form: string | undefined, walletAddress: string, appId: string, logger: LoggerInstance): Promise<Order> {
 
-	const openOrder = await db.OpenOrder.findOneById(orderId);
-	if (!openOrder) {
-		throw Error(`no such order ${orderId}`);
+	const order = await db.Order.findOneById(orderId);
```
Comment (doodyparizada) [inline scripts/src/public/services/orders.ts]: `db.Order.findOne({id: orderId, status: "opened"});`

File: scripts/src/public/services/orders.ts [hunk 6]
```diff
@@ -63,103 +69,114 @@ function orderDbToApi(order: db.Order, logger: LoggerInstance): Order {
...
-	// setTimeout(makeFailed, order.expiration, order.id);
-	return;
+	metrics.submitOrder(offer.type, offer.id);
+	return orderDbToApi(order);
 }
 
 export async function cancelOrder(orderId: string, logger: LoggerInstance): Promise<void> {
 	// you can only delete an open order - not a pending order
-	// validate order
+	const order = await db.Order.findOneById(orderId);
```
Comment (doodyparizada) [inline scripts/src/public/services/orders.ts]: you can only cancel opened:
`db.Order.findOne({id: orderId, status: "opened"});`

File: scripts/src/public/services/orders.ts [hunk 7]
```diff
@@ -63,103 +69,114 @@ function orderDbToApi(order: db.Order, logger: LoggerInstance): Order {
...
+	if (!order) {
+		throw Error(`no such order ${ orderId }`);
+	}
 
-	const openOrder = await db.OpenOrder.findOneById(orderId);
-	if (!openOrder) {
-		throw Error(`no such order ${orderId}`);
+	if (order.status !== "opened") {
```
Comment (doodyparizada) [inline scripts/src/public/services/orders.ts]: instead of this check, move it to the query

File: scripts/src/public/services/orders.ts (hunk 3 above)
Comment (nitzantomer) [reply scripts/src/public/services/orders.ts]: in both cases what i wrote is temporary, we'll need to change it as it's not optimal AT ALL.

**Pull request merged**